x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SetSingleNodeAllocateStep.java

```diff
@@ -69,49 +79,127 @@ public void performAction(IndexMetadata indexMetadata, ClusterState clusterState
+                                shardStats.getStats().getStore().getSizeInBytes()).sum()).sum();
+                        if (shardsOnCurrentNodeStorageBytes != 0) {
+                            nodeShardsStorageBytes.put(nodeId, shardsOnCurrentNodeStorageBytes);
+                        }
+                    }
+                }
+                long indexPrimaryShardsStorageBytes = nodeShardsStorageBytes.values().stream().mapToLong(Long::longValue).sum();
+                if (indexMetadata.getNumberOfReplicas() != 0) {
+                    indexPrimaryShardsStorageBytes /= indexMetadata.getNumberOfReplicas();
+                }
+                for (RoutingNode node : routingNodes) {
+                    if (nodeStatsMap.get(node.nodeId()) != null) {
+                        FsInfo.Path totalFsInfo = nodeStatsMap.get(node.nodeId()).getFs().getTotal();
+                        long nodeTotalBytes = totalFsInfo.getTotal().getBytes();
+                        long nodeAvailableBytes = totalFsInfo.getAvailable().getBytes();
+                        long freeBytesThresholdLow;
+                        if (diskThresholdSettings.getFreeDiskThresholdLow() != 0) {
+                            freeBytesThresholdLow = (long) Math.ceil(nodeTotalBytes *
+                                diskThresholdSettings.getFreeDiskThresholdLow() * 0.01);
+                        } else {
+                            freeBytesThresholdLow = diskThresholdSettings.getFreeBytesThresholdLow().getBytes();
+                        }
+                        // we should make sure that allocating two copies of the index's primary shards to that node doesn't put the node
+                        // above the low watermark, if not, the new shrunken index may not be initialized successfully
+                        long shardsOnCurrentNodeStorageBytes = 0;
+                        if (nodeShardsStorageBytes.containsKey(node.nodeId())) {
+                            shardsOnCurrentNodeStorageBytes = nodeShardsStorageBytes.get(node.nodeId());
+                        }
+                        if (nodeAvailableBytes > freeBytesThresholdLow + 2 * indexPrimaryShardsStorageBytes -
+                            shardsOnCurrentNodeStorageBytes) {
+                            validRoutingNodes.add(node);
+                        }
+                    }
+                }
+
+                for (RoutingNode node : validRoutingNodes) {
+                    boolean canAllocateOneCopyOfEachShard = routingsByShardId.values().stream()
+                        .allMatch(shardRoutings -> shardRoutings.stream()
+                            .map(shardRouting -> allocationDeciders.canAllocate(shardRouting, node, allocation).type())
+                            .anyMatch(Decision.Type.YES::equals));
+                    if (canAllocateOneCopyOfEachShard) {
+                        validNodeIds.add(node.node().getId());
+                    }
+                }
```

Comment:
So I actually think we should do this `canAllocate` check *before* doing the nodes stats call. That way, we can reduce the number of `validNodeIds` that are passed into the nodes stats call, even potentially avoiding the nodes stats call altogether if there are no valid nodes due to a different allocation decider.